rfcs/0107-version-normalization.md

```diff
@@ -0,0 +1,223 @@
+[examples-and-interactions]: #examples-and-interactions
+
+Some useful examples:
+
+- Bochs is a typical Sourceforge-hosted project; its labeled snapshots can be
+  fetched from tarballs obtained via URLs like
+  <https://sourceforge.net/projects/bochs/files/bochs/2.6.11/>
+
+  For this example, we have `pname = "bochs"; version = "2.6.11";`.
+
+- MPV is a typical Github-hosted program; its labeled snapshots can be fetched
+  from tarballs obtained via URLs like
+  <https://github.com/mpv-player/mpv/releases/tag/v0.33.1>.
+
+  For this example, we get rid of the `"v"` prepended to the version tag: `pname
+  = "mpv"; version = "0.33.1";`.
+
+- SDL2 is hosted on Github; its latest labeled version can be downloaded from
+  <https://github.com/libsdl-org/SDL/releases/tag/release-2.0.14>. Therefore we
+  have `pname = "SDL2"; version = "2.0.14";`.
+
+  _However_, this labeled version was released December 21, 2020, while the
+  latest change was done in May 28, 2021.
+
+  Therefore, for this particular unlabeled releases of SDL2, we have `pname =
+  "SDL2"; version = "2.0.14+date=2021-05-28";`.
+
+- Cardboard is a typical Gitlab-hosted program. It has no labeled release yet,
+  therefore we use `0.pre` as default dummy stable version; further, the latest
+  commit was made on May 10, 2021.
+
+  Therefore, for this particular commit have `pname = "cardboard"; version =
+  "0.pre+date=2021-05-21";`.
+
+- Python is a famous programming language and interpreter. Before the
+  deprecation of its 2.x series in 2020, Python had two release branches,
+  popularly known as 'Python 2' and 'Python 3'. Indeed this peculiar situation
+  reflected in many package management teams, especially Nixpkgs, that employed
+  `python2` and `python3` as `pname`s for these particular programs.
+
+  As an exercise of imagination, suppose the scenarios described below:
+
+  Python 2.6 was released 2008-10-01; an unlabeled snapshot of Python 2 branch
+  released at 2008-12-04 would have `version="2.6+date=2008-12-04"`.
+
+  At the same time, Python 3.0 was released 2008-12-03; an unlabeled snapshot of
+  Python 3 branch released at 2008-12-04 would have
+  `version="3.0+date=2008-12-04"`.
+
+- The Linux drivers for Realtek rtl8192eu can be fetched from a Github page,
+  <https://github.com/Mange/rtl8192eu-linux-driver>. It has no labeled release;
+  the latest code is from May 12, 2021. Supposing e.g. it was built for Linux
+  kernel version 5.10.01, we therefore have `pname = "rtl8192eu"; version =
+  "0.pre+date=2021-05-12+linux=5.10.01";`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The main drawback is the conversion of the already existent expressions which
+does not follow the format proposed here, possibly requiring manual intervention
+and code review, especially for machine-generated expressions (such as Lua,
+Emacs Lisp or Node library sets).
+  
+Nonetheless, this task is easily sprintable, can be done incrementally, and is
+amenable to automation.
+
```

Comment:
```suggestion

In some cases, the build system is passed the version, and might be confused by this format. 

However, there are already precedents when Nixpkgs version and build system visible version differ.

```

Reply:
Could it also cause issues with [testVersion](https://nixos.org/manual/nixpkgs/stable/#tester-testVersion)?

Reply:
Probably, although for pre-releases I would expect any use of testVersion to need overrides anyway… Maybe I am too pessimistic.

Reply:
Removing the "R" does not cause issues. It's already done for vapoursynth.
Adding "+date=2021-05-27" causes the build to fail.

```
[davidak@gaming:~/code/nixpkgs]$ nix build .#vapoursynth --no-link --rebuild
error: some outputs of '/nix/store/zi0bpcl0qfdipg0r2pbcd3rpfm7a9kgk-vapoursynth-59+date=2021-05-27.drv' are not valid, so checking is not possible
```

### Workaround

Overwrite the version in the test _should_ solve the issue, but does not.

```
    tests.version = testers.testVersion {
      package = vapoursynth;
      version = "59";
    };
```

```
[davidak@gaming:~/code/nixpkgs]$ nix build .#vapoursynth --no-link --rebuild
error: some outputs of '/nix/store/zi0bpcl0qfdipg0r2pbcd3rpfm7a9kgk-vapoursynth-59+date=2021-05-27.drv' are not valid, so checking is not possible
```

I'm debugging this for 2 hours. I give up for now. Nix is strange.